pkg/reckoner/client.go

```diff
@@ -29,73 +29,86 @@ import (
+	Releases []string
+	// BaseDirectory is the directory where the course file is located
+	BaseDirectory string
+	// DryRun is a flag to indicate if the client should be run in dry run mode
+	DryRun bool
+	// CreateNamespaces is a flag to indicate if the client should create namespaces
 	CreateNamespaces bool
-	ContinueOnError  bool
-	Errors           int
+	// ContinueOnError is a flag to indicate if the client should continue if one release fails
+	ContinueOnError bool
+	// Errors is a counter of errors encountered during the use of this cilent
+	Errors int
+	// HelmArgs is a list of helm args to pass to helm when running commands
+	HelmArgs []string
+	// Schema is a byte slice representation of the coursev2 json schema
+	Schema []byte
 }
 
 var once sync.Once
 var clientset *kubernetes.Clientset
 
-// NewClient returns a client. Attempts to open a v2 schema course file
+// Init initializes a client. Attempts to open a v2 schema course file
 // If getClient is true, attempts to get a Kubernetes client from config
-func NewClient(fileName, version string, plotAll bool, releases []string, kubeClient bool, dryRun bool, createNamespaces bool, schema []byte, continueOnError bool) (*Client, error) {
+func (c *Client) Init(fileName string, initKubeClient bool) error {
 	// Get the course file
-	courseFile, err := course.OpenCourseFile(fileName, schema)
+	courseFile, err := course.OpenCourseFile(fileName, c.Schema)
 	if err != nil {
-		return nil, fmt.Errorf("%w - error opening course file %s: %s", course.SchemaValidationError, fileName, err)
+		return fmt.Errorf("%w - error opening course file %s: %s", course.SchemaValidationError, fileName, err)
 	}
+	c.CourseFile = *courseFile
 
 	// Get a helm client
 	helmClient, err := helm.NewClient()
 	if err != nil {
-		return nil, err
+		return err
 	}
+	c.Helm = *helmClient
 
-	client := &Client{
-		CourseFile:       *courseFile,
-		PlotAll:          plotAll,
-		Releases:         releases,
-		Helm:             *helmClient,
-		ReckonerVersion:  version,
-		BaseDirectory:    path.Dir(fileName),
-		DryRun:           dryRun,
-		CreateNamespaces: createNamespaces,
-		ContinueOnError:  continueOnError,
-	}
+	c.BaseDirectory = path.Dir(fileName)
 
 	// Check versions
-	if !client.helmVersionValid() {
-		return nil, fmt.Errorf("helm version check failed")
+	if !c.helmVersionValid() {
+		return fmt.Errorf("helm version check failed")
+	}
+	if !c.reckonerVersionValid() {
+		return fmt.Errorf("reckoner version check failed")
 	}
-	if !client.reckonerVersionValid() {
-		return nil, fmt.Errorf("reckoner version check failed")
+
+	if err := c.filterReleases(); err != nil {
+		return err
 	}
 
-	if err := client.filterReleases(); err != nil {
-		return nil, err
+	if initKubeClient {
+		c.KubeClient = getKubeClient(courseFile.Context)
 	}
 
-	if kubeClient {
-		client.KubeClient = getKubeClient(courseFile.Context)
+	klog.V(5).Infof("successfully initialized client:")
+	if klog.V(5).Enabled() {
+		spew.Dump(c)
```

Comment:
Is spew used here on purpose, or is this a debug line that should get removed?

Reply:
Since it's gated to log level v5, I figured I would leave it there for future debugging